docs/proposals/local-cluster.md

```diff
@@ -0,0 +1,486 @@
+              |
+              +--- kubectl -> kubectl-1.6.1
+              |      currently used kubectl version (symbolic link)
+              |
+              +--- kubectl-1.6.1-XXXXX
+              |      kubectl binary for k8s 1.6.1 (XXXX is sha1 hash)
+              |
+              +--- kubectl-1.5.4-XXXXX
+                     kubectl binary for k8s 1.5.4 (XXXX is sha1 hash)
+```
+
+The local cluster startup script `local-cluster.sh` works by running
+`local-cluster` docker image, passing all the command line arguments
+and environment variables to the `docker run` command. It sources
+a config file if it exists. The config file path is `~/.local-cluster/config.sh` by default, but this can be overriden using `LOCAL_CLUSTER_CONF` environment variable.
+
+After successful `up` and `reup` operations the script reruns the
+image with `download` command, passing it the list of files in
+`~/.local-cluster`.  The command outputs a compressed tar archive that
+can be empty or can contain the files to put in `.local-cluster/` (the
+proper kubectl binary and a symlink to it, and `admin.conf`). _Why
+this is needed:_ avoid downloading of uncompressed `kubectl` binary if
+the user's machine has slow/limited network connection; reduce the
+amount of logic in the shell script so it doesn't have to be updated
+often; provide an ability to pull extra binaries like `helm` in
+future. Not downloading any binaries isn't an option either because
+e.g. `kubectl` from k8s 1.5 can't properly handle deployments in k8s
+1.6.
+
+When `BUILD_*` options are enabled, `dind-cluster` startup scripts
+invokes appropriate `build/run.sh make ...` commands in Kubernetes
+source tree, preventing the build container from copying back the
+binaries using `KUBE_RUN_COPY_OUTPUT=n`. The control container then
+pulls the binaries directly from build container. _NOTE:_ extra tricks
+may be necessary in case `LOCAL_GO` is enabled.
+
+### Networking
+
+`local-cluster` utilizes named docker network
+`k8s-local-cluster`. This network is always created or replaced upon
+the cluster startup using address range specified via `DIND_SUBNET`
+configuration variable. `local-cluster` makes docker's internal DNS
+server available to all the nodes in the cluster, so nodes can be
+referred by their hostnames: `kube-master`, `kube-node-1`,
+`kube-node-2` and so on.
+
+### Cluster snapshot mechanism
+
+In order to reduce the amount of time needed to restart the cluster,
+`local-cluster` takes the snapshot of the cluster after setting it up.
+The snapshot is made by means of preserving each node's
+`/var/lib/docker` directory as a docker volume and saving each node
+container's filesystem in an image. The volumes and images created by
+snapshotting mechanism are removed during `./local-cluster.sh clean`.
+
+### Future directions
+
+The following features may be added later:
+
+* automatic PV creation like it's done in `minikube`, so most Helm
```

Comment:
this is not easily doable since this is a multinode cluster.

However, we can use Rook or the like easily to get such functionality